### PR TITLE
support multi-message messages

### LINKED ISSUE #41: Item Length Preservation
Currently, the Client will pad Publish messages with zero's up to DataSize. The response will then be the full DataSize long. Do we want to reserve a byte for tagging the actual length of the item?

### PR DESCRIPTION
**Pull request by willscott**
This fixes #41 
It also updates client_test.go to work with new process for generating request vectors introduced in #78

### CODE REVIEW

File: libtalek/client.go [hunk 1]
```diff
@@ -95,30 +96,42 @@ func (c *Client) MaxLength() uint64 {
 func (c *Client) Publish(handle *Topic, data []byte) error {
 	config := c.config.Load().(ClientConfig)
 
-	if len(data) > int(config.DataSize)-PublishingOverhead {
-		return errors.New("message too long")
-	} else if len(data) < int(config.DataSize)-PublishingOverhead {
+	if len(data) > 65535 {
```
Comment (ryscheng) [inline libtalek/client.go]: why 65535? If significant, should be a constant somewhere with a comment explaining

File: libtalek/client.go [hunk 2]
```diff
@@ -95,30 +96,42 @@ func (c *Client) MaxLength() uint64 {
...
+		return errors.New("message is too long")
+	}
+
+	// First word is prepended as length of data:
+	prefixed := make([]byte, 4, 4+len(data))
+	binary.LittleEndian.PutUint32(prefixed[0:4], uint32(len(data)))
+	data = append(prefixed, data...)
+
+	left := len(data)
+	for left > 0 {
```
Comment (ryscheng) [inline libtalek/client.go]: I think the issue with doing it this way is that you're assuming the client gets every single message. In order to support missing a bunch of messages and coming back later, we need to prefix every message with 1. the item length, 2. how many partial messages exist for this message, and 3. which partial message this is. That way we can go offline, come back, and at least know what to ignore to wait for the next full message

File: libtalek/handle.go [hunk 3]
```diff
@@ -142,9 +146,27 @@ func (h *Handle) Decrypt(cyphertext []byte, nonce *[24]byte) ([]byte, error) {
 // sending it to the handle's updates channel if valid.
 func (h *Handle) OnResponse(args *common.ReadArgs, reply *common.ReadReply, dataSize uint) {
 	msg := h.retrieveResponse(args, reply, dataSize)
-	if msg != nil && h.updates != nil {
+	if msg != nil {
 		h.Seqno++
-		h.updates <- msg
+		if h.partialMsgLength == 0 {
```
Comment (ryscheng) [inline libtalek/handle.go]: As mentioned earlier, I think we ought to have some notion of a message header. It'd be cleaner to do this as a struct that gets marshalled to a byte array, than to do this with manual byte manipulation.

File: libtalek/client_test.go [hunk 4]
```diff
@@ -109,13 +110,19 @@ func TestRead(t *testing.T) {
 
 	//Due to thread race, there may be a random read made before
 	//the requested poll is queued up.
-	decRead1, _ := read1.Decode(0, config.TrustDomains[0])
-	decRead2, err := read2.Decode(0, config.TrustDomains[0])
+	decRead11, _ := read1.Decode(0, config.TrustDomains[0])
```
Comment (ryscheng) [inline libtalek/client_test.go]: Why are these changes necessary if there are no new tests?

Comment (willscott) [inline libtalek/client_test.go]: Based on my reading of it, it this broke based on the final change i asked you to make in #78 - when you switched the libtalek to use the common bucket encoding the 1 for the chosen bucket isn't always set in the request to the first trust domain anymore.

File: libtalek/client.go (hunk 1 above)
Comment (willscott) [inline libtalek/client.go]: 2^16 felt reasonable, but we can probably base it off some multiple of the datasize or something? what do you think is reasonable?

File: libtalek/client_test.go [hunk 5]
```diff
@@ -126,3 +126,42 @@ func TestRead(t *testing.T) {
 		t.Fatalf("Read wasn't for the enqueued subscription. %v / %v / %d", rv1, rv2, bucket)
 	}
 }
+
+func TestFragment(t *testing.T) {
+	theMsg := make([]byte, 1024*256)
+	validateMsg := make([]byte, 1024*256)
+
+	hdr := msgHeader{}
```
Comment (ryscheng) [inline libtalek/client_test.go]: Constructor is better so that you don't forget to set left

File: libtalek/client.go [hunk 6]
```diff
@@ -39,6 +39,57 @@ type Client struct {
...
+// Fill will fill a destination buffer with the header, followed by the src msg.
+func (m *msgHeader) Fill(dst, src []byte) int {
+	if len(dst) < 6 {
+		return -1
+	}
+	dst[0] = m.flags
+	binary.LittleEndian.PutUint32(dst[1:5], m.left)
+	num := uint32(copy(dst[5:], src[:]))
+	m.left -= num
+	m.flags |= 1
```
Comment (ryscheng) [inline libtalek/client.go]: If you're going to use a flags byte, it would be a good idea to use helper set/get methods, instead of doing manual byte manipulation every time. This line is hard to parse

File: libtalek/client.go [hunk 7]
```diff
@@ -39,6 +39,57 @@ type Client struct {
...
+}
+
+// Unpack an incoming fragment of message onto a destination buffer.
+// returns whether the destination is now a complete message.
+func (m *msgHeader) Unpack(dest *[]byte, source []byte) (bool, error) {
+	if len(source) < 6 {
+		return false, errors.New("fragment too short")
+	}
+
+	if (m.flags&1) == 0 && (source[0]&1) == 1 {
```
Comment (ryscheng) [inline libtalek/client.go]: Simililarly here, a getter method would help me parse what you're trying to do

File: libtalek/client_test.go [hunk 8]
```diff
@@ -126,3 +126,42 @@ func TestRead(t *testing.T) {
...
+	hdr.left = uint32(len(theMsg))
+
+	parts := make([][]byte, 0)
+	for hdr.left > 0 {
```
Comment (ryscheng) [inline libtalek/client_test.go]: This seems like a really odd abstraction. the msgHeader.Fill is something that the caller needs to iteratively allocate, fill, and advance the message. 

I think what I had in mind, was a Message class. Where a Marshal method would just spit out a array of partial messages and the Unmarshal method would take in all the partials and recompose the message. 

This method seems similarly error prone.
For example, you don't have an ID for the message. If you've missed some messages, you don't know if you're still unpacking the same message or a different one. I'm expecting that we have tests for all of these different failure scenarios. 

When we didn't do multi-message, all these failure scenarios were implicitly covered because messages were self-contained. Now that you're splitting them apart, I'd expect robustness against message drops.

Comment (ryscheng) [inline libtalek/client_test.go]: Also, in theory, it is possible to get partials out of order, if you have a retransmission or something (yea I know thats not implemented yet). So I think you'd need enough state to be able to recompose messages under any ordering and provide enough state to know what to request for a retransmission.

Comment (willscott) [inline libtalek/client_test.go]: uh, i don't think i'm gonna worry about that until we make a protocol where it's possible to happen.  right now it can skip ahead to the next full message if it desynchronizes, which seems like enough.

**Comment by willscott** [comment]
I made a formalized type for messages that encapsulates the split/join logic.

I'm going to argue that we're already over-engineering this component compared to the protocol at this point, and I don't think we can do a particularly good job codifying more about what the fragments look like or need to support in terms of reassembly until we've actually developed what that protocol looks like.

**Comment by ryscheng** [comment]
Looks good thanks for the changes.
I agree we can defer the out of order stuff till later
👍
